src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -999,6 +999,130 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
             return this;
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+
+            if (value != null && value.Length > 0)
+            {
+                if (value != this)
```

Comment:
@AlexRadch while this will guard against the stringbuilder being appended to itself, it won't guard against the case where I am calling `Append` with a `StringBuilder` (let's call it `sb`) from one thread and mutating the state of `sb` from another thread.

The only way I can think of right now (without making changes to the `StringBuilder`) to guard against that is to get a string from the `StringBuilder` and append that. That sadly will have a big perf impact on this API.